### PR TITLE
feat: Return retry-after header in case if rate limit is exceeded #615

### PR DESCRIPTION
**Pull request by astsiapanay**

### CODE REVIEW

File: server/src/main/java/com/epam/aidial/core/server/limiter/RequestRateLimit.java [hunk 1]
```diff
@@ -18,7 +18,10 @@ public RateLimitResult check(long timestamp, Limit limit, long count) {
         if (result) {
             String errorMsg = String.format("Hit request rate limit. Hour limit: %d / %d requests. Day limit: %d / %d requests.",
                     hourTotal, limit.getRequestHour(), dayTotal, limit.getRequestDay());
-            return new RateLimitResult(HttpStatus.TOO_MANY_REQUESTS, errorMsg);
+            long hourRetryAfter = hour.retryAfter(limit.getRequestHour());
+            long dayRetryAfter = day.retryAfter(limit.getRequestDay());
+            long retryAfter = Math.max(hourRetryAfter, dayRetryAfter);
```
Comment (artsiomkorzun) [inline server/src/main/java/com/epam/aidial/core/server/limiter/RequestRateLimit.java:23]: Should you round it towards positive infinity?

File: server/src/main/java/com/epam/aidial/core/server/limiter/RateBucket.java [hunk 2]
```diff
@@ -52,6 +54,22 @@ public long update(long timestamp) {
...
+     */
+    long retryAfter(long limit) {
+        long sum = this.sum;
+        long replyAfter = 0;
+        for (long start = this.start;  start < this.end && sum >= limit; start++) {
+            int index = index(start);
+            sum -= sums[index];
+            replyAfter += window.interval();
+        }
+        return TimeUnit.MILLISECONDS.toSeconds(replyAfter);
```
Comment (artsiomkorzun) [inline server/src/main/java/com/epam/aidial/core/server/limiter/RateBucket.java:70]: Round towards positive infity?

Comment (artsiomkorzun) [reply server/src/main/java/com/epam/aidial/core/server/limiter/RateBucket.java:70]: It always rounded to seconds.

**Review by artsiomkorzun — APPROVED** [review]

**Comment by astsiapanay** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-617.nightly-test.deltixhub.io
>E2E tests status: success

**Pull request merged**